libyul/backends/evm/StackLayoutGenerator.cpp

```diff
@@ -0,0 +1,551 @@
+						},
+						[&](StackSlot const& _s) { return _s == post.at(_target); }
+					}, layout.at(_source))
+				);
+		}
+		bool sourceIsSame(size_t _lhs, size_t _rhs)
+		{
+			return std::visit(util::GenericVisitor{
+				[&](PreviousSlot const&, PreviousSlot const&) { return true; },
+				[&](StackSlot const& _lhs, StackSlot const& _rhs) { return _lhs == _rhs; },
+				[&](auto const&, auto const&) { return false; }
+			}, layout.at(_lhs), layout.at(_rhs));
+		}
+		int sourceMultiplicity(size_t _offset)
+		{
+			return std::visit(util::GenericVisitor{
+				[&](PreviousSlot const&) { return 0; },
+				[&](StackSlot const& _s) { return multiplicity.at(_s); }
+			}, layout.at(_offset));
+		}
+		int targetMultiplicity(size_t _offset)
+		{
+			if (!outputs.count(post.at(_offset)) && !generateSlotOnTheFly(post.at(_offset)))
+				return 0;
+			return multiplicity.at(post.at(_offset));
+		}
+		bool targetIsArbitrary(size_t _offset)
+		{
+			return _offset < post.size() && std::holds_alternative<JunkSlot>(post.at(_offset));
+		}
+		void swap(size_t _i)
+		{
+			yulAssert(!holds_alternative<PreviousSlot>(layout.at(layout.size() - _i - 1)) || !holds_alternative<PreviousSlot>(layout.back()), "");
+			std::swap(layout.at(layout.size() -  _i - 1), layout.back());
+		}
+		size_t sourceSize() { return layout.size(); }
+		size_t targetSize() { return post.size(); }
+		void pop() { layout.pop_back(); }
+		void pushOrDupTarget(size_t _offset) { layout.push_back(post.at(_offset)); }
+	};
+
+	Shuffler<ShuffleOperations>::shuffle(_layout, _post, _generateSlotOnTheFly);
+
+	// Now we can construct the ideal layout before the operation.
+	// "layout" has the declared variables in the desired position and
+	// for any PreviousSlot{x}, x yields the ideal place of the slot before the declaration.
+	vector<optional<StackSlot>> idealLayout(_post.size(), nullopt);
+	for (auto const& [slot, idealPosition]: ranges::zip_view(_post, _layout))
+		if (PreviousSlot* previousSlot = std::get_if<PreviousSlot>(&idealPosition))
+			idealLayout.at(previousSlot->slot) = slot;
+
+	while (!idealLayout.empty() && !idealLayout.back())
+		idealLayout.pop_back();
+
+	return idealLayout | ranges::views::transform([](optional<StackSlot> s) {
+		yulAssert(s, "");
+		return *s;
+	}) | ranges::to<Stack>;
+}
+}
+
+Stack StackLayoutGenerator::propagateStackThroughOperation(Stack _exitStack, CFG::Operation const& _operation)
+{
+	Stack& stack = _exitStack;
+
+	// This is a huge tradeoff between code size, gas cost and stack size.
+	auto generateSlotOnTheFly = [&](StackSlot const&) {
+		//return stack.size() > 12 && canBeFreelyGenerated(_slot);
+		// return canBeFreelyGenerated(_slot);
+		return false;
+	};
+
+	size_t previousLayoutSize = stack.size();
+	for (auto const& slot: stack)
+		if (util::findOffset(_operation.output, slot) || generateSlotOnTheFly(slot))
+				--previousLayoutSize;
+
+	auto layout = ranges::views::iota(0u, previousLayoutSize) |
```

Comment:
"stack" and "layout" are maybe not the best names...

Reply:
I'm really confused now, "stack" and "layout" are two different things and they both get modified in this function?

Reply:
``Layout`` is a symbolic representation of the stack layout immediately after the operation and contains placeholders for variables that stem from before the operation.
It's not really modified, it's just created here and in the next statement. Or well - it is modified in ``createIdealLayout``, but that's an implementation detail of that one - which is indeed missing documentation. ``createIdealLayout`` inverts the stack shuffling algorithm, i.e. it produces a stack that has the layout given by the symbolic layout "_layout", with the placeholders filled in, s.t. shuffling that result to ``_post`` is cheap.

E.g. in
```
let y := 21
let x := 42

x := f(1,2,3)  // This is the operation in question.

// below wants: [x, y]
mstore(x,0)
// below wants: [x]
mstore(y,0)
```
Then ``stack`` is ``[x, y]``.
``x`` is being assigned, so it's an output of the operation. ``y`` is a one "previous slot" that needs to remain unchanged on stack.
So I know that, schematically, the stack before the operation must have the form:
``[PreviousSlot(0), 1, 2, 3]``
then applying the operation yields
``[PreviousSlot(0), x]`` (that's ``layout``)

So I know that after the operation I have ``[PreviousSlot(0), x]``, but I actually *want* to have ``[x,y]``.

So to determine what slot ``PreviousSlot(0)`` is supposed to be, I symbolically shuffle ``layout => stack``, i.e.
``[PreviousSlot(0), x] => [x,y]``, which yields ``[x,PreviousSlot(0)]``. How I can just compare the previous slots with the desired target to see that ``PreviousSlot(0)`` must be ``y``.

Maybe not the best example... gotta run now, I'll try to comment on all of this a bit better in the code and will consider renaming some stuff.

Reply:
But basically the whole thing uses forward-shuffling with symbolic placeholders in the source as a method to invert shuffling.

Reply:
Is the new version I just pushed better? I moved more of it into ``createIdealLayout`` and added some comments about what's happening.